WebKitBrowser/WebKitImplementation.cpp

```diff
@@ -1873,16 +1959,143 @@ static GSourceFuncs _handlerIntervention =
+            } else {
+                TRACE(Trace::Error, (_T("Failed to update environment, host %s:\n"), newWpeHost.c_str()));
+            }
+        }
+
+        std::string getClientCertFromUrl(const std::string& url, std::string& newCertContents)
+        {
+            std::vector<std::string> urlparm_str;
+            string new_url;
+            urlparm_str = tokenize(url, "&?");
+            if(urlparm_str.size() > 0) {
+                static constexpr char payload_cert[] = "clientcert=";
+                static constexpr char payload_certkey[] = "clientcertkey=";
+                static constexpr int  payload_cert_len = sizeof(payload_cert) - 1;
+                static constexpr int  payload_certkey_len = sizeof(payload_certkey) - 1;
+
+                string new_host;
+                string new_payload_cert;
+                string new_payload_certkey;
+                bool useAmpersand = false;
+                for (auto const &urlKeyStr : urlparm_str) {
+                    if (urlKeyStr.compare(0, payload_cert_len, payload_cert) == 0) {
+                        if(new_payload_cert.size() == 0) {
+                            new_payload_cert = urlKeyStr.substr(payload_cert_len);
+                            decodeUrlCertPath(new_payload_cert);
+                            TRACE(Trace::Information, (_T("found parameter clientcert: %s\n"), new_payload_cert.c_str()));
+                        }
+                    } else if ((urlKeyStr.compare(0, payload_certkey_len, payload_certkey)) == 0) {
+                        if(new_payload_certkey.size() == 0) {
+                            new_payload_certkey = urlKeyStr.substr(payload_certkey_len);
+                            decodeUrlCertPath(new_payload_certkey);
+                            TRACE(Trace::Information, (_T("found parameter clientcertkey: %s\n"), new_payload_certkey.c_str()));
+                        }
+                    } else {
+                        if (new_host.size() == 0 ) {
+                            new_host=getUrlOrigin(urlKeyStr);
+                        } else {
+                            if (useAmpersand) {
+                                new_url.append("&");
+                            } else {
+                                useAmpersand = true;
+                                new_url.append("?");
+                            }
+                        }
+                        new_url.append(urlKeyStr);
+                    }
+                }
+                if(new_host.size() && new_payload_cert.size() && new_payload_certkey.size() ) {
+                    AppendClientCertificate(new_host, new_payload_cert, new_payload_certkey, newCertContents);
+                }
+                TRACE_L1("New URL: %s (Base URL: %s)", new_host.c_str(), url.c_str());
+
+            } else {
+                new_url = string(url);
+            }
+            return string(new_url);
+        }
+
+        uint32_t URL(const string& URLNEW) override
```

Comment:
Can we change NEWURL to fullURL or URLwithParams or something like that?
fullURL will indicate that it contains certs, and URL will be stripped fullURL without those additional parameters.

Reply:
updated to: URLwithParams